runelite-client/src/main/java/net/runelite/client/chat/ChatMessageManager.java

```diff
@@ -84,14 +90,392 @@ public void onResizableChanged(ResizeableChanged event)
 		refreshAll();
 	}
 
-	public ChatMessageManager cacheColor(final ChatColor chatColor, final ChatMessageType... types)
+	@Subscribe
+	public void onConfigChanged(ConfigChanged event)
+	{
+		if (event.getGroup().equals("textrecolor"))
+		{
+			loadColors();
+			refreshAll();
+		}
+	}
+
+	@Subscribe
+	public void onSetMessage(SetMessage setMessage)
+	{
+		final Client client = clientProvider.get();
+		MessageNode messageNode = setMessage.getMessageNode();
+		ChatMessageType chatMessageType = setMessage.getType();
+
+		boolean isChatboxTransparent = client.isResized() && client.getVar(Varbits.TRANSPARENT_CHATBOX) == 1;
+		Color usernameColor = null;
+		Color senderColor = null;
+
+		switch (chatMessageType)
+		{
+			case PRIVATE_MESSAGE_RECEIVED_MOD:
+			case PRIVATE_MESSAGE_RECEIVED:
+			case PRIVATE_MESSAGE_SENT:
+				usernameColor = isChatboxTransparent ? chatColorConfig.transparentPrivateUsernames() : chatColorConfig.opaquePrivateUsernames();
+				break;
+
+			case TRADE:
+			case AUTOCHAT:
+			case PUBLIC:
+				usernameColor = isChatboxTransparent ? chatColorConfig.transparentUsername() : chatColorConfig.opaqueUsername();
+				break;
+
+			case CLANCHAT:
+				usernameColor = isChatboxTransparent ? chatColorConfig.transparentClanUsernames() : chatColorConfig.opaqueClanUsernames();
+				break;
+		}
+
+		senderColor = isChatboxTransparent ? chatColorConfig.transparentClanChannelName() : chatColorConfig.opaqueClanChannelName();
+
+		if (usernameColor != null)
+		{
+			messageNode.setName(wrapTextWithColour(messageNode.getName(), usernameColor));
+		}
+
+		String sender = setMessage.getSender();
+		if (senderColor != null && !Strings.isNullOrEmpty(sender))
+		{
+			messageNode.setSender(wrapTextWithColour(sender, senderColor));
+		}
+
+		final Collection<ChatColor> chatColors = colorCache.get(chatMessageType);
+		if (!chatColors.isEmpty())
+		{
+			for (ChatColor chatColor : chatColors)
+			{
+				if (chatColor.isTransparent() != isChatboxTransparent || chatColor.getType() != ChatColorType.NORMAL || chatColor.isDefault())
+				{
+					continue;
+				}
+				messageNode.setValue(wrapTextWithColour(messageNode.getValue(), chatColor.getColor()));
+				break;
+			}
+		}
+	}
+
+	private static String wrapTextWithColour(String text, Color colour)
+	{
+		return "<col=" + Integer.toHexString(colour.getRGB() & 0xFFFFFF) + ">" + text + "</col>";
+	}
+
+	private static Color getDefaultColor(ChatMessageType type, boolean transparent)
+	{
+		if (!transparent)
+		{
+			switch (type)
+			{
+				case PUBLIC:
+					return Color.decode("#0000FF");
```

Comment:
Maybe switching this other way around would reduce the code a bit, eg 1 switch and then return transparent ? : and default: return null